### DISCUSSION TITLE
Zibee2MQTT as a service: Fix for repeated starts/stops/restarts

### DISCUSSION THREAD

**Discussion by ScotsDon** [Koenkk/zigbee2mqtt#21195] [General]
When I followed the Zigbee2MQTT installation instructions on the Zigbee2MQTT website for Linux, I used the offered text for the zigbee2mqtt,service file (/etc/systemd/system/zigbee2mqtt.service).  When I started the service, it ran for a minute or two, then stopped and restarted, continuously.  I had previously tested Zigbee2MQTT on another RPi which I use for testing, and I had not had this problem.  When I compared the two zigbee2mqtt.service files, they were **significantly different** (I had obviouslt used different installation instructions).  So I changed my production Pi to use the same file, and the problem disappeared:  the service now appears to run without these restarts.  I hope this helps anyone with the same problem.

The files are reproduced below, but the main differences are:
**Type is now Type=exec
ExecStart is now: ExecStart=/usr/bin/npm start
WatchdogSec=10s is deleted**

### **ORIGINAL ZIGBEE2MQTT.SERVICE FILE:**
[Unit]
Description=zigbee2mqtt
After=network.target

[Service]
Environment=NODE_ENV=production
Type=notify
ExecStart=/usr/bin/node index.js
WorkingDirectory=/opt/zigbee2mqtt
'#StandardOutput=inherit
'# Or use StandardOutput=null if you don't want Zigbee2MQTT messages filling syslog, for more options see systemd.exec(5)
StandardOutput=null
StandardError=inherit
WatchdogSec=10s
Restart=always
RestartSec=10s
User=pi

[Install]
WantedBy=multi-user.target

### **REVISED ZIGBEE2MQTT.SERVICE FILE:**
[Unit]
Description=zigbee2mqtt
After=network.target

[Service]
Environment=NODE_ENV=production
Type=exec
ExecStart=/usr/bin/npm start
WorkingDirectory=/opt/zigbee2mqtt
'#StandardOutput=inherit
StandardOutput=null
'# Or use StandardOutput=null if you don't want Zigbee2MQTT messages filling syslog, for more options see systemd.exec(5)
StandardError=inherit
Restart=always
RestartSec=10s
User=pi

[Install]
WantedBy=multi-user.target

**Comment by Nardol** [comment]
If you run `sudo apt install libsystemd-dev` you will be able to use the new service which is in the installation guide.
I'm going to make a pull request to add this in the packages to install.

Pull request opened, see Koenkk/zigbee2mqtt.io#2538

**Comment by ScotsDon** [comment]
Sounds like a good idea, but I'll wait until it's all done.  At the moment. it works.

**Comment by Nardol** [comment]
> Sounds like a good idea, but I'll wait until it's all done. At the moment. it works.

The solution I give is the one I applied when I had the same issue you describe, I just procrastinated too much and forgot to make the PR I made after reading your message :wink: but it is up to you of course :slightly_smiling_face:

**Comment by ScotsDon** [comment]
I've spent quite a long time doing this, by no means all of it on this particular issue.  So I intend to stop while I'm on top!  I can control Zigbee devices from Python now, which is what I wanted to do.  However, it would be sensible for me to upgrade.  Am I right in assuming that the installation instructions were updated before the latest software was released?  So upon release, the "original" zigbee2mqtt.service file will work.  If so, and not wanting to deviate from the production release, how will I know when this occurs?

FYI, I am a bit of a newbie with zigbee,, although I have used ZWave for many years.  I am a private hobby developer and have created a number of things for my house.  I am not very familiar with Github.

Thanks for your help and advice.

**Reply by Nardol** [reply]
In this specific case, documentation has been updated after the release, but there was a missing point, the one I made a pull request for today.
The SystemD unit file you use is working because exec mode does not require any implementation in Zigbee2MQTT itself.
But Notify requires a module (sd-notify) which needs libsystemd-dev to be built.
Changing to notify is not required at all, in fact we are both right.
Notify allows SystemD to restart Zigbee2MQTT automatically in the case it has not notified SystemD it is still running, it should be safer than previous method IMO.

I hop I explained all you needed and was enough clear in my explanations :slightly_smiling_face:

**Reply by ScotsDon** [reply]
Thanks, I think I understand it.  Would you mind me asking what your role is?

**Reply by Nardol** [reply]
I am a Zigbee2MQTT user, like you are.
